Switch to Webpack

andrewcoelho opened the pull request:
Works alright for the most part. This is a WIP.

TODO:
- [x] Production config
- [x] Combine dev and prod configs into one
- [x] Add necessary scripts for linting, building, etc.
- [x] Find a way to make template reloading work

(CircleCI currently fails because of unmet peer dependencies caused by beta version of webpack)

@endtwist please review

andrewcoelho committed:
Remove Gulp and Browserify and switch to Webpack

andrewcoelho committed:
Add and modify webpack configs and clean package.json

Adds a basic production config and modifies the existing
dev config to cache the Babel build folder. This allows
faster rebuilding if the assets haven't changed.

package.json was also cleaned up to remove unnecessary
dependencies.

andrewcoelho committed:
Merge separate webpack config and update package.json

Merges the separate production and development webpack
configs into one. To do this in a more understandable
and concise way, a few packages were installed, namely
webpack-config-utils. In addition to this, webpack config
validation was added as well as a lightweight progress
bar to reportstatus of builds.

Additionally, the package.json was updated to include
new build scripts for production and development. Previously,
webpack-dev-server ran using the node.js api, which has been
swapped out in favor of CLI usage.

Unnecessary packages were also stripped away to keep dependencies
down to the essentials.

andrewcoelho committed:
Add long-term asset caching and NPM scripts

Generation of a manifest file was finalized to be inlined
in the head of the outputted html. This adds a content-dependent
cache-buster to each file. The entry point chunk containing the
bootstrapping code doesn't change its hash over time if its
dependencies are unchanged.

A copy plugin was also added to the Webpack config which moves
fonts into the build folder. Styles and scripts are moved by
Webpack after they have been processed.

The imagemin-cli and additional NPM build scripts were added to
facilitate image minification and relocation to the build directory.

Finally, the ESLint config was updated to use the new eslint-config-planetary
package for simplification of linting rule management.

andrewcoelho committed:
Switch from SCSS to PostCSS

Replaces scss in favor of postcss. A few plugins have been
added to PostCSS to enable usage of next-gen CSS features.
The main workhorse here is cssnext, which allows color
manipulation, nesting, and a variety of other useful features
from the future css spec.

A few additional postcss plugins were added to provide additional
functionality. For the initial transition to postcss, the goal
was to keep additional plugins beyond cssnext to a minimum.

The plugins which were chosen for inclusion were postcss-mixins,
postcss-modular-scale, and postcss-import. While the names are
somewhat self-explanatory, there are a few quirks with modular-scale.

The modular-scale plugin does not seem to appropriately output
the correct sizes at this point, as the --ms-bases and --ms-ratios
does not have any effect. The outputted scale is modular, but the
ratio used appears to be 1.618 instead of the inputted 1.5.

Yet to be added is a stylelint config.

andrewcoelho committed:
Fix modular scale and add stylelint config

A previous commit used a postcss plugin for achieving
modular scales which did not respond to changes in base
size or ratios. This plugin was replaced in favor
of the native modular scale css implementation which
behaves appropriately.

Additionally, a stylelint config was added as well as
a new build plugin for webpack to lint everything in
the assets/styles directory except for vendor stylesheets.

ilikescience committed:
bring layout file up-to-date, remove normalize.css in favor of custom reset

andrewcoelho committed:
Enable auto refresh on template changes

A webpack plugin was developed as a separate
npm package, reload-html-webpack-plugin. This plugin
injects a small script into the body of the html outputted
by html-webpack-plugin. This script receives socket events
emitted by html-webpack-plugin and refreshes the browser
accordingly.

ilikescience committed:
add stylelintrc, update files to reflect rules

andrewcoelho committed:
Merge branch 'enhancement/webpackify' into enhancement/postcss

andrewcoelho committed:
Fix typo in media.css

andrewcoelho committed:
Merge branch 'enhancement/postcss' into enhancement/webpackify

andrewcoelho committed:
Update readme to reflect switch to PostCSS

andrewcoelho committed:
Update babel-eslint to v6

andrewcoelho committed:
Downgrade to webpack v1

Webpack is downgraded to v1 while the official
release of webpack v2 drops. This also means removing
the reload-html-webpack-plugin which only works with webpack v2.

andrewcoelho committed:
Reimplement automatic browser refresh

andrewcoelho committed:
Update circle.yml

endtwist commented on webpack.config.babel.js:
```diff
@@ -0,0 +1,124 @@
+                test: /\.(jpe?g|png|gif|svg)$/i,
+                loaders: [
+                    'file?name=images/[name].[ext]',
+                    'image-webpack?bypassOnDebug&optimizationLevel=7&interlaced=false'
```
We should consider inlining small images (< 10kb?) in the CSS, too.

andrewcoelho committed:
Add inlining of images under 10kb and fix package.json

Addition of url loader to automatically inline images as a base64
that are smaller than 10kb.

The postbuild scripts for imagemin were also updated to output
images to the correct directory in build.

andrewcoelho committed:
Add automatic loading of view templates

Adds a custom function to iterate over each .pug file
in the assets/templates/views directory and automatically
creates a new instance of html-webpack-plugin for each one.

This function is loaded in the plugins section of the webpack
config using a destructuring of the array returned from loadHtmlPlugin.

endtwist commented on webpack.config.babel.js:
```diff
@@ -106,13 +125,7 @@ module.exports = webpackValidator({
-            template: './templates/views/styles.pug',
-            filename: 'styles.html'
-        }),
+        ...loadHtmlPlugin(),
```
Nice!
